pageserver/src/metrics.rs

```diff
@@ -1223,29 +1223,41 @@ pub(crate) mod virtual_file_io_engine {
     });
 }
 
-pub(crate) struct SmgrOpTimer {
+pub(crate) struct SmgrOpTimer(Option<SmgrOpTimerInner>);
+pub(crate) struct SmgrOpTimerInner {
     global_latency_histo: Histogram,
 
     // Optional because not all op types are tracked per-timeline
     per_timeline_latency_histo: Option<Histogram>,
 
+    /// Always Some() except when we transition into [`SmgrOpFlushInProgress`]
+    global_flush_in_progress_micros: Option<IntCounter>,
+    /// See [`Self::global_flush_in_progress_micros`].
+    per_timeline_flush_in_progress_micros: Option<IntCounter>,
+
     start: Instant,
     throttled: Duration,
     op: SmgrQueryType,
 }
 
-impl SmgrOpTimer {
+pub(crate) struct SmgrOpFlushInProgress {
+    base: Instant,
+    global_micros: IntCounter,
+    per_timeline_micros: IntCounter,
+}
+
+impl SmgrOpTimerInner {
     pub(crate) fn deduct_throttle(&mut self, throttle: &Option<Duration>) {
         let Some(throttle) = throttle else {
             return;
         };
         self.throttled += *throttle;
     }
-}
 
-impl Drop for SmgrOpTimer {
-    fn drop(&mut self) {
-        let elapsed = self.start.elapsed();
+    /// Must only call once!
+    fn smgr_op_end(&self) -> Instant {
```

Comment:
nit: can we assert on this with an `ended` member?

Reply:
Addressed in recent pushes, please check

Reply:
looks good - thanks for tweaking